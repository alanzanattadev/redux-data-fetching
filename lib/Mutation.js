// @flow

export default class Mutation {
  mutationQL: string;
  variables: { [variableName: string]: mixed };
  onError: () => void;
  onCompleted: () => void;
  operationName: ?string;

  constructor(
    {
      mutationQL,
      variables = {},
      onError = () => {},
      onCompleted = () => {},
      operationName,
    }: {
      mutationQL: string,
      variables?: { [variableName: string]: mixed },
      onError?: () => void,
      onCompleted?: () => void,
      operationName?: string,
    } = {},
  ) {
    if (mutationQL == null) {
      throw new Error(
        "mutationQL has to be defined in the params of new Mutation: new Mutation({mutationQL: `mutation ...`})",
      );
    }
    this.mutationQL = mutationQL;
    this.variables = variables;
    this.onError = onError;
    this.onCompleted = onCompleted;
    this.operationName = operationName;
  }

  setOperationName(name: string) {
    return new Mutation({
      mutationQL: this.mutationQL,
      variables: this.variables,
      onError: this.onError,
      onCompleted: this.onCompleted,
      operationName: name,
    });
  }
}
