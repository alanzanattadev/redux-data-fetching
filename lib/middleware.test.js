// @flow

import {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLString,
  GraphQLList,
} from "graphql";
import configureMiddleware from "./middleware";
import configureActions from "./actions";
import Mutation from "./Mutation";

describe("Middleware", () => {
  const User = new GraphQLObjectType({
    name: "User",
    fields: () => ({
      id: { type: GraphQLString },
      name: { type: GraphQLString },
      description: { type: GraphQLString },
      friends: { type: new GraphQLList(User) },
    }),
  });
  const schema = new GraphQLSchema({
    query: new GraphQLObjectType({
      name: "RootQuery",
      fields: {
        user: {
          type: User,
          resolve: () => ({ id: "1", name: "Alan" }),
          args: { id: { type: GraphQLString } },
        },
        users: {
          type: new GraphQLList(User),
          resolve: (p, args, context) => [],
        },
      },
    }),
    mutation: new GraphQLObjectType({
      name: "RootMutation",
      fields: {
        createUser: {
          type: User,
          args: {
            name: { type: GraphQLString },
          },
          resolve: () => ({ id: "1", name: "Alan" }),
        },
      },
    }),
  });
  const actions = configureActions();
  const normalizrModel = {
    converters: {
      user: "User",
      users: "Users",
      createUser: "User",
    },
  };
  const context = {};
  const middleware = configureMiddleware(
    schema,
    actions,
    normalizrModel,
    context,
  );

  describe("GRAPHQL_FETCH", () => {
    it("dispatches queryStarted and packageData", done => {
      let calls = 0;
      const dispatchSpy = jest.fn(() => {
        if (calls === 1) {
          expect(dispatchSpy.mock.calls).toMatchSnapshot();
          done();
        }
        calls = calls + 1;
      });
      const nextSpy = jest.fn();
      const subject = middleware({ getState: () => {}, dispatch: dispatchSpy })(
        nextSpy,
      )({
        type: "GRAPHQL_FETCH",
        graphql: true,
        payload: `{
        user (id: "1") {
          id,
          name
        }
      }`,
      });
    });

    it("dispatches queryStarted and queryFailed if query fails", done => {
      let calls = 0;
      const dispatchSpy = jest.fn(() => {
        if (calls === 1) {
          expect(dispatchSpy.mock.calls).toMatchSnapshot();
          done();
        }
        calls = calls + 1;
      });
      const nextSpy = jest.fn();
      const subject = middleware({ getState: () => {}, dispatch: dispatchSpy })(
        nextSpy,
      )({
        type: "GRAPHQL_FETCH",
        graphql: true,
        payload: `{
        user (badargument: "1") {
          id,
          name
        }
      }`,
      });
    });
  });

  describe("GRAPHQL_MUTATION", () => {
    it("doesn't error with variables", done => {
      let calls = 0;
      const dispatchSpy = jest.fn(() => {
        if (calls === 1) {
          expect(dispatchSpy.mock.calls[1][0].type).toMatchSnapshot();
          done();
        }
        calls = calls + 1;
      });
      const nextSpy = jest.fn();
      const subject = middleware({ getState: () => {}, dispatch: dispatchSpy })(
        nextSpy,
      )({
        type: "GRAPHQL_MUTATION",
        graphql: true,
        payload: {
          queryID: "id",
          mutation: new Mutation({
            mutationQL: `mutation CreateUser ($name: String) {
              createUser (name: $name) {
              id,
              name
            }
          }`,
            variables: {
              name: "Alan",
            },
          }),
        },
      });
    });

    it("doesn't error without variables", done => {
      let calls = 0;
      const dispatchSpy = jest.fn(() => {
        if (calls === 1) {
          expect(dispatchSpy.mock.calls[1][0].type).toMatchSnapshot();
          done();
        }
        calls = calls + 1;
      });
      const nextSpy = jest.fn();
      const subject = middleware({ getState: () => {}, dispatch: dispatchSpy })(
        nextSpy,
      )({
        type: "GRAPHQL_MUTATION",
        graphql: true,
        payload: {
          queryID: "id",
          mutation: new Mutation({
            mutationQL: `mutation CreateUser {
              createUser {
              id,
              name
            }
          }`,
          }),
        },
      });
    });

    it("doesn't error without variables and without name", done => {
      let calls = 0;
      const dispatchSpy = jest.fn(() => {
        if (calls === 1) {
          expect(dispatchSpy.mock.calls[1][0].type).toMatchSnapshot();
          done();
        }
        calls = calls + 1;
      });
      const nextSpy = jest.fn();
      const subject = middleware({ getState: () => {}, dispatch: dispatchSpy })(
        nextSpy,
      )({
        type: "GRAPHQL_MUTATION",
        graphql: true,
        payload: {
          queryID: "id",
          mutation: new Mutation({
            mutationQL: `mutation {
              createUser {
              id,
              name
            }
          }`,
          }),
        },
      });
    });

    it("dispatches queryStarted and packageData", done => {
      let calls = 0;
      const dispatchSpy = jest.fn(() => {
        if (calls === 1) {
          expect(dispatchSpy.mock.calls).toMatchSnapshot();
          done();
        }
        calls = calls + 1;
      });
      const nextSpy = jest.fn();
      const subject = middleware({ getState: () => {}, dispatch: dispatchSpy })(
        nextSpy,
      )({
        type: "GRAPHQL_MUTATION",
        graphql: true,
        payload: {
          queryID: "id",
          mutation: new Mutation({
            mutationQL: `mutation CreateUser {
              createUser {
              id,
              name
            }
          }`,
          }),
        },
      });
    });

    it("calls onCompleted", done => {
      const onCompletedSpy = jest.fn(param => {
        expect(param).toMatchSnapshot();
        done();
      });
      const nextSpy = jest.fn();
      const subject = middleware({ getState: () => {}, dispatch: () => {} })(
        nextSpy,
      )({
        type: "GRAPHQL_MUTATION",
        graphql: true,
        payload: {
          queryID: "id",
          mutation: new Mutation({
            mutationQL: `mutation CreateUser {
              createUser {
              id,
              name
            }
          }`,
            onCompleted: onCompletedSpy,
          }),
        },
      });
    });

    it("dispatches queryStarted and queryFailed if query fails", done => {
      let calls = 0;
      const dispatchSpy = jest.fn(() => {
        if (calls === 1) {
          expect(dispatchSpy.mock.calls).toMatchSnapshot();
          done();
        }
        calls = calls + 1;
      });
      const nextSpy = jest.fn();
      const subject = middleware({ getState: () => {}, dispatch: dispatchSpy })(
        nextSpy,
      )({
        type: "GRAPHQL_MUTATION",
        graphql: true,
        payload: {
          queryID: "id",
          mutation: new Mutation({
            mutationQL: `mutation CreateUser($input: User) {
              createUser (badarg: 2) {
              id,
              name
            }
          }`,
          }),
        },
      });
    });

    it("calls onError if query fails", done => {
      const onErrorSpy = jest.fn(param => {
        expect(param).toMatchSnapshot();
        done();
      });
      const subject = middleware({
        getState: () => {},
        dispatch: () => {},
      })(() => {})({
        type: "GRAPHQL_MUTATION",
        graphql: true,
        payload: {
          queryID: "id",
          mutation: new Mutation({
            mutationQL: `mutation CreateUser($input: User) {
              createUser (badarg: 2) {
              id,
              name
            }
          }`,
            onError: onErrorSpy,
          }),
        },
      });
    });
  });

  describe("General", () => {
    it("calls next if it's not a handled action", () => {
      const spy = jest.fn();
      const dispatch = jest.fn();

      middleware({ getState: () => {}, dispatch })(spy)({
        type: "CREATE_USER",
      });
      middleware({ getState: () => {}, dispatch })(spy)({
        type: "GRAPHQL_FETCH",
      });
      middleware({ getState: () => {}, dispatch })(spy)({
        type: "GRAPHQL_MUTATION",
      });

      expect(spy).toHaveBeenCalledTimes(3);
      expect(dispatch).not.toHaveBeenCalled();
    });
  });
});
