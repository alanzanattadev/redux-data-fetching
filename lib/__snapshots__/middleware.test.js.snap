// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Middleware GRAPHQL_FETCH dispatches queryStarted and packageData 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "query": Object {
          "request": Object {
            "hash": 398953408,
            "ql": "{
        user (id: \\"1\\") {
          id,
          name
        }
      }",
          },
        },
      },
      "type": "QUERY_STARTED",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "entities": Object {
          "User": Array [
            Object {
              "id": "1",
              "name": "Alan",
            },
          ],
        },
        "query": Object {
          "request": Object {
            "hash": 398953408,
            "ql": "{
        user (id: \\"1\\") {
          id,
          name
        }
      }",
          },
          "response": Object {
            "raw": Object {
              "user": Object {
                "id": "1",
                "name": "Alan",
              },
            },
          },
        },
      },
      "type": "DATA_RECEIVED",
    },
  ],
]
`;

exports[`Middleware GRAPHQL_FETCH dispatches queryStarted and packageData with a Query 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "query": Object {
          "request": Object {
            "hash": 1686497088,
            "ql": "{
            user (id: \\"1\\") {
              id,
              name
            }
          }",
          },
        },
      },
      "type": "QUERY_STARTED",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "entities": Object {
          "User": Array [
            Object {
              "id": "1",
              "name": "Alan",
            },
          ],
        },
        "query": Object {
          "request": Object {
            "hash": 1686497088,
            "ql": "{
            user (id: \\"1\\") {
              id,
              name
            }
          }",
          },
          "response": Object {
            "raw": Object {
              "user": Object {
                "id": "1",
                "name": "Alan",
              },
            },
          },
        },
      },
      "type": "DATA_RECEIVED",
    },
  ],
]
`;

exports[`Middleware GRAPHQL_FETCH dispatches queryStarted and queryFailed if query fails 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "query": Object {
          "request": Object {
            "hash": 259749885,
            "ql": "{
        user (badargument: \\"1\\") {
          id,
          name
        }
      }",
          },
        },
      },
      "type": "QUERY_STARTED",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "errors": Array [
          [GraphQLError: Unknown argument "badargument" on field "user" of type "RootQuery".],
        ],
        "query": Object {
          "request": Object {
            "hash": 259749885,
            "ql": "{
        user (badargument: \\"1\\") {
          id,
          name
        }
      }",
          },
        },
      },
      "type": "QUERY_FAILED",
    },
  ],
]
`;

exports[`Middleware GRAPHQL_MUTATION calls onCompleted 1`] = `undefined`;

exports[`Middleware GRAPHQL_MUTATION calls onError if query fails 1`] = `undefined`;

exports[`Middleware GRAPHQL_MUTATION dispatches queryStarted and packageData 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "query": Object {
          "request": Object {
            "hash": "id",
            "ql": "mutation CreateUser {
              createUser {
              id,
              name
            }
          }",
          },
        },
      },
      "type": "QUERY_STARTED",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "entities": Object {
          "User": Array [
            Object {
              "id": "1",
              "name": "Alan",
            },
          ],
        },
        "query": Object {
          "request": Object {
            "hash": "id",
            "ql": "mutation CreateUser {
              createUser {
              id,
              name
            }
          }",
          },
          "response": Object {
            "raw": Object {
              "createUser": Object {
                "id": "1",
                "name": "Alan",
              },
            },
          },
        },
      },
      "type": "DATA_RECEIVED",
    },
  ],
]
`;

exports[`Middleware GRAPHQL_MUTATION dispatches queryStarted and queryFailed if query fails 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "query": Object {
          "request": Object {
            "hash": "id",
            "ql": "mutation CreateUser($input: User) {
              createUser (badarg: 2) {
              id,
              name
            }
          }",
          },
        },
      },
      "type": "QUERY_STARTED",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "errors": Array [
          [GraphQLError: Variable "$input" cannot be non-input type "User".],
          [GraphQLError: Unknown argument "badarg" on field "createUser" of type "RootMutation".],
          [GraphQLError: Variable "$input" is never used in operation "CreateUser".],
        ],
        "query": Object {
          "request": Object {
            "hash": "id",
            "ql": "mutation CreateUser($input: User) {
              createUser (badarg: 2) {
              id,
              name
            }
          }",
          },
        },
      },
      "type": "QUERY_FAILED",
    },
  ],
]
`;

exports[`Middleware GRAPHQL_MUTATION doesn't error with variables 1`] = `"DATA_RECEIVED"`;

exports[`Middleware GRAPHQL_MUTATION doesn't error without variables 1`] = `"DATA_RECEIVED"`;

exports[`Middleware GRAPHQL_MUTATION doesn't error without variables and without name 1`] = `"DATA_RECEIVED"`;
